dattri/algorithm/data_shapley.py

```diff
@@ -0,0 +1,205 @@
+                should not be shuffled.
+            train_labels: (List[int], optional): The list of training labels,
+                with the same size and order of the training dataset. If not provided,
+                the last element in each batch from the loader will be used as label.
+            test_labels: (List[int], optional): The list of test labels,
+                with the same size and order of the test dataset. If not provided,
+                the last element in each batch from the loader will be used as label.
+
+        Returns:
+            Tensor: The shapley values of the training set on the test set, with
+                the shape of (num_train_samples, num_test_samples).
+
+        Raises:
+            ValueError: If the length of provided labels and dataset mismatch.
+        """
+        _check_shuffle(test_dataloader)
+        _check_shuffle(train_dataloader)
+
+        if (train_labels is not None and
+            len(train_labels) != len(train_dataloader.sampler)):
+            error_msg = "Length of training labels and training dataset mismatch."
+            raise ValueError(error_msg)
+
+        if (test_labels is not None and
+            len(test_labels) != len(test_dataloader.sampler)):
+            error_msg = "Length of test labels and test dataset mismatch."
+            raise ValueError(error_msg)
+
+        num_train_samples = len(train_dataloader.sampler)
+        num_test_samples = len(test_dataloader.sampler)
+
+        dist_matrix = torch.zeros(
+            size=(num_test_samples, num_train_samples),
+        )
+
+        shapley_values = torch.zeros(
+            size=(num_test_samples, num_train_samples),
+        )
+
+        if train_labels is None:
+            train_labels = []
+
+        if test_labels is None:
+            test_labels = []
+
+        for test_batch_idx, test_batch_data in enumerate(
+            tqdm(
+                test_dataloader,
+                desc="fitting KNN...",
+                leave=False,
+            ),
+        ):
+            for train_batch_idx, train_batch_data in enumerate(
+                train_dataloader,
+            ):
+                partial_dist = self.distance_func(test_batch_data, train_batch_data)
+
+                # results position based on batch info
+                col_st = train_batch_idx * train_dataloader.batch_size
+                col_ed = min(
+                    (train_batch_idx + 1) * train_dataloader.batch_size,
+                    len(train_dataloader.sampler),
+                )
+
+                row_st = test_batch_idx * test_dataloader.batch_size
+                row_ed = min(
+                    (test_batch_idx + 1) * test_dataloader.batch_size,
+                    len(test_dataloader.sampler),
+                )
+
+                dist_matrix[row_st:row_ed, col_st:col_ed] += partial_dist
+
+                if len(train_labels) != len(train_dataloader.sampler):
```

Comment:
Add a comment here that we are assuming train_batch_data is a tuple with the last element as the batch of train_labels

Reply:
Done.